packages/checkout/widgets-lib/src/lib/hooks/useBalance.ts

```diff
@@ -0,0 +1,78 @@
+import { useCallback, useEffect, useState } from 'react';
+import { Checkout, GetBalanceResult } from '@imtbl/checkout-sdk';
+import { Web3Provider } from '@ethersproject/providers';
+import { getTokenBalances } from '../../widgets/wallet/functions/tokenBalances';
+import { DEFAULT_BALANCE_RETRY_POLICY } from '../constants';
+import { useInterval } from './useInterval';
+
+const REFRESH_BALANCE_INTERVAL_MS = 10000;
+
+export interface UseBalanceParams {
+  checkout: Checkout | undefined;
+  provider: Web3Provider | undefined,
+  refreshCallback: (balance: GetBalanceResult[]) => void;
+  errorCallback: (error: Error) => void;
+}
+
+export const useBalance = ({
+  checkout,
+  provider,
+  refreshCallback,
+  errorCallback,
+}: UseBalanceParams) => {
+  const [balancesLoading, setBalancesLoading] = useState(true);
+
+  const refreshBalances = useCallback(async (
+    // If silent is true, the balances will not be set to loading
+    silent: boolean = false,
+  ) => {
+    if (!checkout || !provider) return;
+    try {
+      const network = await checkout.getNetworkInfo({
+        provider,
+      });
+
+      /* If the provider's network is not supported, return out of this and let the
+      connect loader handle the switch network functionality */
+      if (!network.isSupported) {
+        return;
+      }
+
+      if (!silent) {
+        setBalancesLoading(true);
+      }
+      const balances: GetBalanceResult[] = await getTokenBalances(checkout, provider, network.chainId);
+      if (!silent) {
+        setBalancesLoading(false);
+      }
+      refreshCallback(balances);
+
+      // Ignore errors given that this is a background refresh
+      // and the logic will retry anyways.
+    } catch (error: any) {
+      if (DEFAULT_BALANCE_RETRY_POLICY.nonRetryable!(error)) {
+        errorCallback(error);
+      }
+      if (!silent) {
+        setBalancesLoading(false);
+      }
+    }
+  }, [checkout, provider]);
+  useInterval(() => refreshBalances(true), REFRESH_BALANCE_INTERVAL_MS);
+
+  // Listen for window focus event to refresh after a potential interaction with a wallet
+  useEffect(() => {
+    const handleFocus = () => {
+      refreshBalances(true);
+    };
+    window.addEventListener('focus', handleFocus);
+    return () => {
+      window.removeEventListener('focus', handleFocus);
```

Comment:
Is it possible to pass in a ref to the widget rather than attaching the listener to the window? I am thinking, what happens if the widget is simply brought back to screen by a user click, would that trigger a refresh?

Reply:
@andrearampin it depends on how the widget is hidden from view by the developer. 
- If it's using the mount/unmount via our API it will trigger a refresh. 
- If the user clicks anywhere outside of the browser window (including browser extensions) it will trigger a refresh when coming back (focus event).
- If they are hiding the UI with CSS then it wont trigger a refresh until 10sec idle
- Some other events like Smart Checkout buy will flush the balance cache as well